app/javascript/crayons/Button/Button.jsx

```diff
@@ -0,0 +1,110 @@
+import { h } from 'preact';
+import PropTypes from 'prop-types';
+import { defaultChildrenPropTypes } from '../../src/components/common-prop-types';
+
+function getAdditionalClassNames({
+  variant,
+  className,
+  icon,
+  disabled,
+  children,
+}) {
+  let additionalClassNames = '';
+
+  if (variant && variant.length > 0 && variant !== 'primary') {
+    additionalClassNames += ` crayons-btn--${variant}`;
+  }
+
+  if (icon) {
+    additionalClassNames +=
+      children.length > 0
+        ? ' crayons-btn--icon-left'
+        : ' crayons-btn--icon-alone';
+  }
+
+  if (disabled) {
+    additionalClassNames += ' crayons-btn--disabled';
+  }
+
+  if (className && className.length > 0) {
+    additionalClassNames += ` ${className}`;
+  }
+
+  return additionalClassNames;
+}
+
+export const Button = ({
+  children,
+  variant = 'primary',
+  as = 'button',
+  className,
+  icon,
+  url,
+  buttonType,
+  disabled,
+  onClick,
+  onMouseOver,
+  onMouseOut,
+  onFocus,
+  onBlur,
+}) => {
+  const ComponentName = as;
+  const Icon = icon;
+  const otherProps =
+    as === 'button'
+      ? { type: buttonType, disabled }
+      : { href: disabled ? undefined : url };
+
+  return (
+    <ComponentName
+      className={`crayons-btn${getAdditionalClassNames({
+        variant,
+        className,
+        icon,
+        disabled: as === 'a' && disabled,
```

Comment:
`htmlElement === a` looks cool

Reply:
Instead of `htmlElement`, it actually makes sense to call it `tagName` as that's what it really is, https://developer.mozilla.org/en-US/docs/Web/API/Element/tagName.

Reply:
Great! `tagName` which follows HTML conventions!

Reply:
Done.